Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43926 [Wip] Rewrote Adadelta with _foreach APIs
* #43925 [wip] Rewrote adamax optimizer with _foreach APIs
* #43727 WIP asgd
* #43726 [wip] Rewrote rprop optimizer with _foreach APIs
* #43725 [wip] Rewrote rmsprop with _forerach APIs
* #43724 [wip] Rewrote SGD with _foreach APIs
* #43562 [wip] Rewrote AdamW with _foreach APIs
* #43507 [WIP] Rewrote adam optimizer with foreach APIs
* #43413 Added alpha overloads for add/sub ops with lists
* #42760 Added pointwise ops: addcdiv, addcmul
* #42537 Add unary ops: exp and sqrt
* #42536 Add binary ops for _foreach APIs
* **#42533 Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs**

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331894](https://our.internmc.facebook.com/intern/diff/D23331894)

Commit:
Added add_list APIs

[ghstack-poisoned]

Commit:
Update on "[IGNORE] Added add_list APIs"




[ghstack-poisoned]

Commit:
Update on "[IGNORE] Added add_list APIs"




[ghstack-poisoned]

Commit:
Update on "[IGNORE] Added add_list APIs"




[ghstack-poisoned]

Commit:
Update on "[IGNORE] Added add_list APIs"




[ghstack-poisoned]

Commit:
Update on "[IGNORE] Added add_list APIs"




[ghstack-poisoned]

Commit:
Update on "[IGNORE] Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"




[ghstack-poisoned]

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

gchanan commented:
which optimizer needs this?

ngimel commented:
out of place list add e.g. sgd `d_p = d_p.add(buf, alpha=momentum)`, in-place list add adam and adamw: `exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)`

gchanan commented:
broadcasting?

ngimel commented:
`exp_avg` and `grad` have the same size. If broadcasting happens due to `grad` being expanded (that should not be common) then fast path won't be taken.

ezyang commented:
Assuming the PRs follow the model of the previous one I probably won't have anything substantive to say. Tag me if there's something specific you want me to look at.

gchanan commented:
> exp_avg and grad have the same size. If broadcasting happens due to grad being expanded (that should not be common) then fast path won't be taken.

I'd expect the decision to broadcast or not to at least be included in the description of the PR.

gchanan commented on aten/src/ATen/native/cuda/ForeachTensorAddList.cu:
```diff
@@ -0,0 +1,205 @@
+
+namespace {
+
+template<typename x_t, typename y_t>
```
why are there multiple scalar_types if that's not supported?

Commit:
Update on "[WIP]  Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP]  Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/cuda/ForeachTensorAddList.cu:
```diff
@@ -0,0 +1,205 @@
+namespace {
+
+template<typename x_t, typename y_t>
+struct AddListFunctor_
```
I think it's worth splitting these Functors with a slim interface into another file to separate the wrapping code from the implementation (which might change in the future). This seems a bit too specific to multi_tensor_apply.

Commit:
Update on "[WIP]  Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP]  Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP]  Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP]  Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list

2. Rewrite PyTorch optimizers to use for-each operators in order to get performance gains.

[ghstack-poisoned]

Commit:
Update on "[WIP]  Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

[ghstack-poisoned]

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

[ghstack-poisoned]

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

[ghstack-poisoned]

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -52,4 +53,49 @@ bool check_fast_route(TensorList tensors, Scalar scalar) {
+  for (int i = 0; i < tensors1.size(); i++) {
+    TORCH_CHECK(tensors1[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
+    TORCH_CHECK(tensors2[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
+    TORCH_CHECK(tensors1[i].sizes() == tensors2[i].sizes(), "Corresponding tensors in lists must have the same size.");
```
what about strides?

cpuhrsch replied:
Ok those don't matter for the slow route

cpuhrsch commented on test/test_foreach.py:
```diff
@@ -119,6 +119,52 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+            torch._foreach_add(tensors1, tensors2)
+            torch._foreach_add_(tensors1, tensors2)
+
+        # Different dtypes
```
What about different devices?

izdeby replied:
device is a fast vs slow path check so it will require a benchmark test to test

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331894](https://our.internmc.facebook.com/intern/diff/D23331894)

[ghstack-poisoned]

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331894](https://our.internmc.facebook.com/intern/diff/D23331894)

[ghstack-poisoned]

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +119,52 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+        tensors2 = []
+
+        # Empty lists
+        with self.assertRaises(RuntimeError):
+            torch._foreach_add(tensors1, tensors2)
+            torch._foreach_add_(tensors1, tensors2)
+
```
`self.assertRaises` will catch the RuntimeError the first time it happens in the block, so we aren't actually testing the second line for if it raises a regex or not.

This should be something like:

```
with self.assertRaises(RuntimeError):
    torch._foreach_add(tensors1, tensors2)
with self.assertRaises(RuntimeError):
    torch._foreach_add_(tensors1, tensors2)
```
so that we test that both raise RuntimeErrors, not just the first one

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +119,52 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+        tensors2 = []
+
+        # Empty lists
+        with self.assertRaises(RuntimeError):
```
nit: prefer self.assertRaisesRegex so that we can check that the error message is sane and makes sense

zou3519 replied:
You can just use a part of the string as the regex. E.g., "self.assertRaisesRegex(RuntimeError, "There were no tensor arguments to this function")"

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +119,52 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+        with self.assertRaises(RuntimeError):
+            torch._foreach_add(tensors1, tensors2)
+            torch._foreach_add_(tensors1, tensors2)
+
```
There should also be a test that "all tensors in the list must have the same size"

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +119,52 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+            torch._foreach_add_(tensors1, tensors2)
+
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_add_list_same_size(self, device, dtype):
```
Maybe add a test for non-contiguous tensors?

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +119,52 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_add_list_same_size(self, device, dtype):
+        tensors1 = []
+        tensors2 = []
+        for _ in range(10):
+            tensors1.append(torch.zeros(10, 10, device=device, dtype=dtype))
+            tensors2.append(torch.ones(10, 10, device=device, dtype=dtype))
+
+        res = torch._foreach_add(tensors1, tensors2)
+        torch._foreach_add_(tensors1, tensors2)
+        self.assertEqual(res, tensors1)
+        self.assertEqual(res[0], torch.ones(10, 10, device=device, dtype=dtype))
```
According to your PR message (I haven't read through the implementation), there exists a "fast path" and a "slow path". We should try to test both paths by writing test cases that invoke both. The code in this test case looks like it only invokes the fast path.

zou3519 replied:
EDIT: it looks like CPU always goes through fast route, and CUDA sometimes goes through the fast route or the slow route. What we're currently testing in this PR:
- CPU slow route
- CUDA fast route

What is missing:
- CUDA slow route test

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +119,52 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+        torch._foreach_add_(tensors1, tensors2)
+        self.assertEqual(res, tensors1)
+        self.assertEqual(res[0], torch.ones(10, 10, device=device, dtype=dtype))
+
```
It would be nice to test some edge cases (e.g., number of tensors in the list >= 100).

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +119,52 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+        tensors1 = []
+        tensors2 = []
+        for _ in range(10):
+            tensors1.append(torch.zeros(10, 10, device=device, dtype=dtype))
+            tensors2.append(torch.ones(10, 10, device=device, dtype=dtype))
```
Would like to see some tests for CUDA tensors that are not on the default device. I'm a bit suspicious of the "device_guard: False" statements in native_functions.yaml, but I haven't read through the entire thing yet

zou3519 replied:
I see, `multi_tensor_apply` comes with its own device guard

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -54,5 +54,48 @@ bool check_fast_route(TensorList tensors, Scalar scalar) {
 
-} // namespace
+void verify_list(TensorList tensors1, TensorList tensors2) {
+  TORCH_CHECK(tensors1.size() > 0, "Tensor list must have at least one tensor.");
+  TORCH_CHECK(tensors2.size() > 0, "Tensor list must have at least one tensor.");
+  TORCH_CHECK(tensors1.size() == tensors2.size(), "Tensor lists must have the same number of tensors.");
+
+  auto expected_dtype = tensors1[0].dtype();
+
+  for (int i = 0; i < tensors1.size(); i++) {
+    TORCH_CHECK(tensors1[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
+    TORCH_CHECK(tensors2[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
+    TORCH_CHECK(tensors1[i].sizes() == tensors2[i].sizes(), "Corresponding tensors in lists must have the same size.");
```
nit: It's kind of nice to report to the user what quantities existed in their code so that they're easier to debug. e.g. "Tensor lists must have the same number of tensors, got 3 and 5"

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5737,6 +5737,20 @@
+    CPU: foreach_tensor_add_list_kernel_slow
+    CUDA: foreach_tensor_add_list_kernel_cuda
+
+- func: _foreach_add_.List(Tensor[](a!) self, Tensor[] other) -> ()
```
Shouldn't this be `Tensor(a!)[]` ? I'm reading `Tensor[](a!)` as "the list self is being modified in-place", and "Tensor(a!)[]" as "the list self is not being modified, but its constituent tensors are being modified in-place"

zou3519 commented on aten/src/ATen/native/ForeachOpsKernels.cpp:
```diff
@@ -21,4 +22,24 @@ void foreach_tensor_add_scalar_kernel_slow_(TensorList tensors, Scalar scalar) {
+  std::vector<Tensor> result;
+  for (int i = 0; i < tensors1.size(); i++) {
+    auto temp = tensors1[i].add(tensors2[i]);
+    result.emplace_back(temp);
```
nit: this should be `result.emplace_back(std::move(temp))`, to avoid an extra refcount bump on Tensor. (example demonstrating that this code, without std::move, copies `temp`: https://godbolt.org/z/5reb1T)

cpuhrsch replied:
Or you can remove temp and just write ```result.push_back(tensors1[i].add(tensors2[i]))``` https://godbolt.org/z/M3qade

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -54,5 +54,48 @@ bool check_fast_route(TensorList tensors, Scalar scalar) {
+
+  auto expected_dtype = tensors1[0].dtype();
+
+  for (int i = 0; i < tensors1.size(); i++) {
```
The PR body implies that "All tensors in the list must have the same [device]". Should that be checked here?

zou3519 commented on aten/src/ATen/native/cuda/ForeachTensorAddList.cu:
```diff
@@ -0,0 +1,47 @@
+std::vector<Tensor> foreach_tensor_add_list_kernel_cuda(TensorList tensors1, TensorList tensors2) {
+    verify_list(tensors1, tensors2);
+
+    if (!check_fast_route(tensors1, tensors2)) {
```
nit: As a function name, "check_fast_route" doesn't sound like it would return a boolean (it sounds like it would return nothing). Something like "can_use_fast_route" sounds like it would return a boolean.

zou3519 commented on aten/src/ATen/native/cuda/ForeachTensorAddList.cu:
```diff
@@ -0,0 +1,47 @@
+        return at::native::foreach_tensor_add_list_kernel_slow(tensors1, tensors2);
+    }
+
+    std::vector<std::vector<at::Tensor>> tensor_lists; 
```
(no action required) We know exactly how many lists there needs to be, so it would be nice if we could avoid the dynamic allocation here. This would require us to revisit the first argument to `multi_tensor_apply`.

zou3519 commented on aten/src/ATen/native/cuda/ForeachTensorAddList.cu:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    std::vector<std::vector<at::Tensor>> tensor_lists; 
+    std::vector<at::Tensor> vec_res;
```
I've been told it is good practice to reserve space for vectors beforehand to avoid the vector re-sizing itself as elements are added. See https://www.geeksforgeeks.org/using-stdvectorreserve-whenever-possible/.

This would look like:
```
std::vector<at::Tensor> vec_res;
vec_res.reserve(tensors1.size());
```

zou3519 commented on aten/src/ATen/native/cuda/ForeachFunctors.cuh:
```diff
@@ -121,6 +121,134 @@ struct AddScalarFunctor {
+                }
+            }
+            else {
+                // Non-divergent exit condition for __syncthreads, not necessary here
```
Where is the `__syncthreads` this is referring to? What exactly is "not necessary here"?

zou3519 commented on aten/src/ATen/native/cuda/ForeachFunctors.cuh:
```diff
@@ -121,6 +121,134 @@ struct AddScalarFunctor {
+
+            n -= chunk_idx * chunk_size;
+
+            T r_x[kILP];
```
What is kILP? It looks like it is equal to 4, but I'm not sure what that means

zou3519 commented on aten/src/ATen/native/cuda/ForeachFunctors.cuh:
```diff
@@ -121,6 +121,134 @@ struct AddScalarFunctor {
+        TensorListMetadata<3>& tl) {
+            int tensor_loc = tl.block_to_tensor[blockIdx.x];
+            int chunk_idx = tl.block_to_chunk[blockIdx.x];
+            int n = tl.sizes[tensor_loc];
```
Probably would have preferred "tl.numel_for_tensor" instead of "tl.sizes", it took me a few minutes of hunting around to figure out that "sizes" refers to numel and not the actual shape of the Tensor

zou3519 commented on aten/src/ATen/native/cuda/ForeachFunctors.cuh:
```diff
@@ -121,6 +121,134 @@ struct AddScalarFunctor {
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_out[ii] = static_cast<T>(r_x[ii]) + static_cast<T>(r_y[ii]);
```
It seems like the important part of the Functor is `r_out[ii] = static_cast<T>(r_x[ii]) + static_cast<T>(r_y[ii]);` and that the other logic is pretty standard between Functors (from me comparing AddListFunctor and AddListFunctor_). Do we ever want functors that don't have similar logic? If not, it might be worth trying to abstract away the rest of the code via templates or macros so that someone implementing a new Functor doesn't have to copy & paste code that is considered to be boilerplate.

zou3519 replied:
(to clarify, no action necessary, but I am curious about if we can clean this up in the future)

zou3519 replied:
I'm also not sure what the functors will look like in the end state so I agree it makes sense to refactor later

zou3519 reviewed (commented):
I need to go read through how multi_tensor_apply and functors work, but the code looks like it is following in the footsteps of the last PRs so I don't have any major comments. Added some minor comments about the testing and some questions in-line

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331894](https://our.internmc.facebook.com/intern/diff/D23331894)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachOpsKernels.cpp:
```diff
@@ -21,4 +22,23 @@ void foreach_tensor_add_scalar_kernel_slow_(TensorList tensors, Scalar scalar) {
+std::vector<Tensor> foreach_tensor_add_list_kernel_slow(TensorList tensors1, TensorList tensors2) {
+  verify_list(tensors1, tensors2);
+
+  std::vector<Tensor> result;
```
nit: `result.reserve(tensors1.size())`

izdeby replied:
Done

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -9,16 +9,31 @@ void verify_list(TensorList tensors) {
 }
 
-// To go via 'fast' path, several conditions must be satisfied 
+void verify_list(TensorList tensors1, TensorList tensors2) {
```
A comment here that these are the restrictions for the foreach TensorList APIs would be nice

izdeby replied:
Done

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +120,96 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+        # different devices
+        tensor1 = torch.zeros(10, 10, device="cuda:0")
+        tensor2 = torch.ones(10, 10, device="cuda:1")
+        with self.assertRaisesRegex(RuntimeError, "Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1!"):
```
Feel free to just check part of the string, e.g. "Expected all tensors to be on the same device". The goal of checking assertRaisesRegex is to make sure the error message isn't something completely unreadable like "TORCH_INTERNAL_ASSERT(false)".

zou3519 commented on test/test_foreach.py:
```diff
@@ -119,6 +120,96 @@ def test_add_scalar_with_different_tensor_dtypes(self, device):
+        # Coresponding tensors with different sizes 
+        tensors1 = [torch.zeros(10, 10, device=device) for _ in range(10)]
+        tensors2 = [torch.ones(11, 11, device=device) for _ in range(10)]
+        with self.assertRaisesRegex(RuntimeError, "Corresponding tensors in lists must have the same size, got \[10, 10\] and \[11, 11\]"):
```
`'\['` isn't a valid escape sequence for regular strings, you have to append an r to the front of the string:
```
r"Corresponding tensors in lists..."
```

izdeby replied:
Done

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -9,16 +9,31 @@ void verify_list(TensorList tensors) {
 }
 
-// To go via 'fast' path, several conditions must be satisfied 
+void verify_list(TensorList tensors1, TensorList tensors2) {
```
nit: this is in the at::native namespace. To avoid name collision, it might be nice to name this something more detailed, like "check_foreach_api_restrictions"

izdeby replied:
Done

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331894](https://our.internmc.facebook.com/intern/diff/D23331894)

[ghstack-poisoned]

Commit:
Update on "Add  _foreach_add(TensorList tl1, TensorList tl2) and _foreach_add_(TensorList tl1, TensorList tl2) APIs"


[First PR: Add private API to support tensor lists: _foreach_add(TensorList tensors, Scalar scalar)](https://github.com/pytorch/pytorch/pull/41554).

**Motivation**
[GitHub issue](https://github.com/pytorch/pytorch/issues/38655) 
Current PyTorch optimizer implementations are not efficient in cases when we work with a lot of small feature tensors. Starting a lot of kernels slows down the whole process. We need to reduce the number of kernels that we start. 
As an example, we should be looking at [NVIDIAs Apex](https://github.com/NVIDIA/apex). 
In order to track progress, we will pick PyTorchs DCGAN model with Adam optimizer and once the optimizer is reimplemented with tensor lists, benchmark the model performance against original model version, Apexs version with original Adam optimizer and it’s FusedAdam optimizer.

**Current API restrictions**
- List can't be empty (will fixed in upcoming PRs). 
- All tensors in the list must have the same dtype, device and size.

**Broadcasting**
At this point we don't support broadcasting. 

**What is 'Fast' and 'Slow' route**
In particular cases, we cant process an op with a fast list CUDA kernel. Still, we can do with a regular for-loop where the op will be applied to each tensor individually through the dispatch mechanisms. There are a few checks that decide whether the op will be performed via a 'fast' or 'slow' path. 
To go the fast route,
- All tensors must have strided layout
- All tensors must be dense and not have overlapping memory
- The resulting tensor type must be the same. 

----------------
**In this PR**
- Adding a `_foreach_add(TensorList tl1, TensorList tl2)` API
- Adding a `_foreach_add_(TensorList tl1, TensorList tl2)` API

**Tests**
Tested via unit tests

**TODO**
1. Properly handle empty lists

**Plan for the next PRs**
1. APIs
- Binary Ops for list with Scalar
- Binary Ops for list with list
- Unary Ops for list
- Pointwise Ops

2. Complete tasks from TODO
3. Rewrite PyTorch optimizers to use for-each operators for performance gains.

Differential Revision: [D23331894](https://our.internmc.facebook.com/intern/diff/D23331894)

[ghstack-poisoned]
